### ISSUE TITLE
[libcxx] Lookup array size in _BMSkipTable for std::boyer_moore_horspool_searcher is not sufficient

### ISSUE THREAD

**Issue by ibmibmibm** [llvm/llvm-project#63822]
In `include/experimental/functionall:174`:
```c++
//  Special case small numeric values; use an array
template<class _Key, typename _Value, class _Hash, class _BinaryPredicate>
class _BMSkipTable<_Key, _Value, _Hash, _BinaryPredicate, true> {
private:
    typedef _Value value_type;
    typedef _Key   key_type;

    typedef typename make_unsigned<key_type>::type unsigned_key_type;
    typedef std::array<value_type, numeric_limits<unsigned_key_type>::max()> skip_map;
    skip_map __table;
```
The `key_type` evaluate to `char`, thus the `numeric_limits<>::max()` evaluate to 255, which is less-by-one for correct array size 256.

**Comment by philnik777** [comment]
Do you have a test case that fails/crashes?

**Comment by ibmibmibm** [comment]
```c++
#include <cstdio>
#include <experimental/functional>
#include <string>

using namespace std::literals;

int main(int argc, char *argv[]) {
  const auto Pattern = "a"s;
  std::experimental::boyer_moore_horspool_searcher Searcher(Pattern.begin(), Pattern.end());
  const auto Corpus = "\xff\xff"s;
  auto Iter = std::search(Corpus.begin(), Corpus.end(), Searcher);
  std::printf("%ld\n", std::distance(Corpus.begin(), Iter));
  return 0;
}
```

```bash
root@92c153c51f51:~# clang++ --version
clang version 15.0.0 (https://github.com/llvm/llvm-project.git bf7f8d6fa6f460bf0a16ffec319cd71592216bf4)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/local/bin
root@92c153c51f51:~# clang++ test.cpp -std=c++17 -Og -fsanitize=address -stdlib=libc++
root@92c153c51f51:~# ./a.out
=================================================================
==234==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x61d000000890 at pc 0x0000004dc180 bp 0x7fffd6a7cf30 sp 0x7fffd6a7cf28
READ of size 8 at 0x61d000000890 thread T0
    #0 0x4dc17f in main (/root/a.out+0x4dc17f)
    #1 0x7fb97f1aa082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #2 0x41d39d in _start (/root/a.out+0x41d39d)

0x61d000000890 is located 0 bytes to the right of 2064-byte region [0x61d000000080,0x61d000000890)
allocated by thread T0 here:
    #0 0x4d961d in operator new(unsigned long) /src/llvm-project/compiler-rt/lib/asan/asan_new_delete.cpp:95:3
    #1 0x4dbd2e in main (/root/a.out+0x4dbd2e)
    #2 0x7fb97f1aa082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)

SUMMARY: AddressSanitizer: heap-buffer-overflow (/root/a.out+0x4dc17f) in main
Shadow bytes around the buggy address:
  0x0c3a7fff80c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c3a7fff80d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c3a7fff80e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c3a7fff80f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c3a7fff8100: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c3a7fff8110: 00 00[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fff8120: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fff8130: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fff8140: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fff8150: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fff8160: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==234==ABORTING

```

**Comment by philnik777** [comment]
This has been fixed in 971e9c8, which should be in LLVM 15. I guess you just have to update your toolchain.
